lib/ex_unit/lib/ex_unit/formatter.ex

```diff
@@ -200,25 +202,101 @@ defmodule ExUnit.Formatter do
     formatter.(:error_info, value)
   end
 
-  defp code_multiline(expr, padding) when is_binary(expr) do
-    String.replace(expr, "\n", "\n" <> String.duplicate(" ", padding))
+  defp code_multiline(expr, padding_size) when is_binary(expr) do
+    padding = String.duplicate(" ", padding_size)
+    String.replace(expr, "\n", "\n" <> padding)
   end
 
-  defp code_multiline(expr, padding) do
-    code_multiline(expr |> Macro.to_string, padding)
+  defp code_multiline(expr, padding_size) do
+    code_multiline(Macro.to_string(expr), padding_size)
   end
 
-  defp inspect_multiline(expr, padding, width) do
-    width = if width == :infinity, do: width, else: width - padding
-    expr
-    |> inspect(pretty: true, width: width)
-    |> String.replace("\n", "\n" <> String.duplicate(" ", padding))
+  defp inspect_multiline(expr, padding_size, width) do
+    padding = String.duplicate(" ", padding_size)
+    width = if width == :infinity, do: width, else: width - padding_size
+    inspect(expr, [pretty: true, width: width])
+    |> String.replace("\n", "\n" <> padding)
   end
 
   defp make_into_lines(reasons, padding) do
     padding <> Enum.join(reasons, "\n" <> padding) <> "\n"
   end
 
+  defp format_diff(struct, formatter) do
+    if_value(struct.left, fn left ->
+      if_value(struct.right, fn right ->
+        if same_data_type?(left, right) do
+          format_diff(left, right, formatter)
+        else
+          ExUnit.AssertionError.no_value
+        end
+      end)
+    end)
+  end
+
+  defp format_diff(left, right, formatter) when is_binary(left) and is_binary(right) do
+    String.difference(left, right)
+    |> Enum.map_join(&format_diff_fragment(&1, formatter))
+    |> String.replace("\n", "\n" <> @inspect_padding)
+  end
+
+  defp format_diff(left, right, formatter) when is_map(left) and is_map(right) do
+    {surplus, altered} =
+      Enum.reduce(left, {[], []}, fn({key, val1}, {surplus, altered} = acc) ->
+        case Map.fetch(right, key) do
+          {:ok, ^val1} ->
+            acc
+          {:ok, val2} ->
+            {surplus, [{key, {val1, val2}} | altered]}
+          :error ->
+            {[{key, val1} | surplus], altered}
+        end
+      end)
+    missing = Enum.reject(right, fn {key, _} -> Map.has_key?(left, key) end)
+    result = Enum.reduce(missing, [], fn({key, val}, acc) ->
+      [formatter.(:diff_insert, inspect(key) <> " => " <> inspect(val)) | acc]
+    end)
+    result = Enum.reduce(surplus, result, fn({key, val}, acc) ->
+      [formatter.(:diff_delete, inspect(key) <> " => " <> inspect(val)) | acc]
+    end)
+    result = Enum.reduce(altered, result, fn({key, {val1, val2}}, acc) ->
```

Comment:
Maybe we should use the regular difference here. I.e. call format_diff/3 recursively.